### ISSUE TITLE
Reconsider having a Default Topology

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.AzureServiceBus#181]
Connects to Particular/PlatformDevelopment#718

#### Context

- The implicit topology for the ASB transport versions 6 and below was a topology identical to the `EndpointOrientedTopology` (there was no concept of a Topology before)
- The default topology for the ASB transport version 7 is the `EndpointOrientedTopology`
- 100% of existing users that will be upgrading, will be on the `EndpointOrientedTopology`
- For the new users, the `ForwardingTopology` will not be an obvious option due to an implicit default set to the `EndpointOrientedTopology`

#### Concerns

1. Should the default topology be set implicitly and which one?
2. How to make the new `ForwardingTopology` visible for the new users that do not know they have a choice?

Currently, `EndpointOrientedTopology` is the **default** topology set by the transport. This will lead to a confusion (as described [here](https://github.com/Particular/NServiceBus.AzureServiceBus/issues/177) ). Which begs a question: should the ASB transport set a default topology at all? Since publishers for the `EndpointOrientedTopology` **have** to be defined, what's the benefit of making it a default one? 

In addition to that, by having a default topology set to the `EndpointOrientedTopology` hides the option of using an alternative topology (`ForwardingTopology`) from the new users, leading them to use a topology w/o considering pros and cons.

By **removing the default topology** we'll achieve the following:

1. Existing customers will still be specifying the publishers, _but_ they will be aware of the new topology that they could try to upgrade to if they are willing to do so.
1. New users will be forced to select a topology, and by doing so, will be encouraged to read the documentation. [The documentation](https://github.com/Particular/docs.particular.net/pull/1411) will guide the new users to prefer the `ForwardingTopology` over `EndpointOrientedTopology`.

**Comment by SeanFeldman** [comment]
@ramonsmits @timbussmann @yvesgoeleven @andreasohlund - fyi

**Comment by timbussmann** [comment]
:+1: to make the choice explicit. What about the topology namings and api discoverability? `.UseTopology<T>()` is hard to discover from a user perspective. Why not `.UseLegacyTopology` and `.UseForwardingTopology()`? I think renaming it to something like LegacyTopology will help in new users not selecting it and existing users realize that they should consider switching to the new topology at some point.

**Comment by SeanFeldman** [comment]
@timbussmann we have considered to call `EndpointOrientedTopology` a "legacy" topology. And have realized 2 things: 
1. It's not a legacy when 100% of our upgrading users are on it
2. Even `ForwardingTopology` can turn into a `LegacyTopology`

**Comment by SeanFeldman** [comment]
@timbussmann nitpicking here, but 

```c#
.UseTopology<T>()
```

and 

```c#
.UseLegacyTopology()
```

would have the same discoverability issue :stuck_out_tongue: I'd rather go for a solid doco in this case.

**Comment by timbussmann** [comment]
@SeanFeldman no they have not the same discoverability. The second approach will easily list all available topologies in intellisense while this doesn't work as smooth with the generic parameter approach.

> It's not a legacy when 100% of our upgrading users are on it

It's not the number of users which decide this? At some point we no longer want to support it and recommend a different topology, so it becomes a legacy component with a new major version.

With the current naming, there is no way for a user to be able to select a topology without consulting the docs first, that seems to be a pain point to me.

**Comment by SeanFeldman** [comment]
@timbussmann I argue that having no default topology will force customers to select and read about differences. For existing customers, I suspect upgrade won't be something they'll jump into. For the new users that are not biased, we'll guide through documentation and outlined pros/cons of the topologies.

**Comment by mikeminutillo** [comment]
```CSharp
.UseToplogy<T>()
```

is consistent with `UseSerializer<>`, `UseTransport<>`, and `UsePersistence<>`.

I also would rather not name something `LegacyX`. It's better if we name them after what they do and then customers can decide which one makes sense for them. This is a new concept. It makes sense to consult the docs just as it would have when Persistence was first introduced.

**Comment by melkio** [comment]
Using specific methods to express the choice, like `UseEndpointOrientedTopology()` and `UseForwardingTopology()`, we could add `publishers config API` to the first method as parameter, enforcing the needs to configure them for `EndpointOrientedTopology`

In this case we have to add a "general" method to support custom topologies too. Something like `UseCustomTopology<T>()`

**Comment by yvesgoeleven** [comment]
```CSharp
.UseToplogy<T>()
```

This is consistent with other API's, it's extensible. We alse considered names like StandardTopology, LegacyTopology, but we didn't like it. The topology is not legacy, it served us very well for years and can serve existing customers for years to come. It has some shortcomings that we can solve since the introduction of the Forwarding feature, but that doesn't mean it's bad. Furthermore, introductions of future features in SB will probably lead to even better ways to set it up, resulting in 2 legacy topologies.

**Comment by yvesgoeleven** [comment]
@melkio we cannot force the configuration of publisher mappings, not all endpoints will subscribe, it's an optional thing by definition

**Comment by melkio** [comment]
@yvesgoeleven introducing `UseEndpointOrientedTopology(x => mappingHere)` (or something similar) a user could provide an "empty" lambda, if their endpoint doesn't need to subscribe.
In this way we can guide users through the needs to configure `publishers`

**Comment by yvesgoeleven** [comment]
@melkio as I used I do not get exited about empty required lambda's

**Comment by melkio** [comment]
@yvesgoeleven we can provide an overload without parameters, for endpoint that doesn't need to subscribe.

My point is: with current implementation, config API doesn't highlight the needs to this _new_ mapping to users. Using a specific method, as proposed before, we can provide a better guide through the possible options.
f.i.: @ramon updated transport package and configured all the things. Endpoint failed at runtime because he didn't know anything about the needs to map publishers. 
The issue here is that config API didn't suggest to him to provide mapping between publisher names and event types.

**Comment by timbussmann** [comment]
> @yvesgoeleven we can provide an overload without parameters, for endpoint that doesn't need to subscribe.

then again we can't make sure users know that they should configure publisher names? Seems like an empty lambda would be the necessary evil in case you're not handling events.

Since `UnicastBusConfig` is public, would it be possible to provide an `mappings => mappings.ReadFromConfigFile()` extension which accesses the config section?

**Comment by yvesgoeleven** [comment]
@melkio once you provide the overload it's not more discoverable then an extension method, both require you to do an additional step in VS to see that there is more to configure (check the overload, or check the follow up methods in intellisene)

@timbussmann we no longer have access to the core api's that allow us to figure out the publishers from those mappings, we need the message metadata repository for that, which needs to be resolved from the container, which we don't have access to

**Comment by melkio** [comment]
@yvesgoeleven IMO it's not the same. When you select `UseEndpointOrientedTopology()` method, intellisense shows to users that the method has two overloads. With an extension method a user has to write an additional `.` to go through others available methods.

**Comment by melkio** [comment]
>  which needs to be resolved from the container, which we don't have access to

This is another thing that I don't understand. Core has a container and transport(s) can't access it. Transports are extensions to core, but they belong to the same framework. why is there a so-clear split between core and transports?

**Comment by SeanFeldman** [comment]
Let's get back to the original concerns of this issue (issue description).
Publishers API is a different issue, #179.

By removing the default topology, do we achieve the following:

> How to make the new `ForwardingTopology` visible for the new users that do not know they have a choice?

And the answer is yes. It is consistent with other [core-related configuration APIs](https://github.com/Particular/NServiceBus.AzureServiceBus/issues/181#issuecomment-209790318). Users (new or old) will be aware of the new concept of `Topology` introduced by the transport and will encourage them to look up the documentation for more information.

**Comment by yvesgoeleven** [comment]
@SeanFeldman yes, remove the default.

**Comment by SeanFeldman** [comment]
Fixed by 2ee943cc472171bafb0534f9b14cc97193c1f4ea
